### PR TITLE
WS c++ -> rust simpler implementation

### PR DESCRIPTION
**Pull request by sinkingsugar**

### CODE REVIEW

File: rust/src/shards/ws.rs [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+    // we need to keep this alive here, otherwise it will be dropped at the end of this function
+    self.ws = Rc::new(Some(ws.0));
+    let var_ws = Var::new_object(&self.ws, &WS_CLIENT_TYPE);
+    Ok(var_ws)
+  }
+}
+
+#[derive(Default)]
+struct ClientUser {
+  ws: Rc<Option<WsClient>>,
```
Comment (guusw) [inline rust/src/shards/ws.rs:86]: Why not `Option<Rc<WsClient>>` or even just `Option<WsClient>`? It doesn't need to stay alive past the lifetime of the `Client` shard I believe.

File: rust/src/shards/ws.rs [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+    self.client.warmup(context)
+  }
+
+  fn cleanup(&mut self) -> Result<(), &str> {
+    self.client.cleanup()
+  }
+
+  fn activate(&mut self, _context: &Context, input: &Var) -> Result<Var, &str> {
+    // TODO make this run on a worker thread but not using BlockingShard as would overlap/crash
+    let ws = self.client.activate()?;
```
Comment (guusw) [inline rust/src/shards/ws.rs:305]: Where is the overlap, is it in the websocket library itself?

File: rust/src/shards/ws.rs (hunk 1 above)
Comment (sinkingsugar) [reply rust/src/shards/ws.rs:86]: yeah has to be that way so the client could even disappear

File: rust/src/shards/ws.rs (hunk 2 above)
Comment (sinkingsugar) [reply rust/src/shards/ws.rs:305]: no it was here, both were using await

**Pull request merged**